backend/de.metas.manufacturing/src/main/java/org/eevolution/productioncandidate/service/produce/PPOrderProducerFromCandidate.java

```diff
@@ -0,0 +1,171 @@
+import java.util.stream.Stream;
+
+public class PPOrderProducerFromCandidate
+{
+	private final OrderGenerateResult result;
+	private final PPOrderAllocatorService ppOrderAllocatorBuilderService;
+	private final PPOrderCandidateDAO ppOrderCandidatesDAO;
+	private final IPPOrderBL ppOrderService;
+	private final ITrxManager trxManager;
+	private final IProductPlanningDAO productPlanningsRepo;
+	private final Map<ProductPlanningId, I_PP_Product_Planning> planning2RecordShortTimeIndex;
+
+	@Builder
+	public PPOrderProducerFromCandidate(
+			@NonNull final PPOrderAllocatorService ppOrderAllocatorBuilderService,
+			@NonNull final IPPOrderBL ppOrderService,
+			@NonNull final ITrxManager trxManager,
+			@NonNull final PPOrderCandidateDAO ppOrderCandidatesDAO,
+			@NonNull final IProductPlanningDAO productPlanningsRepo)
+	{
+		this.ppOrderAllocatorBuilderService = ppOrderAllocatorBuilderService;
+		this.ppOrderService = ppOrderService;
+		this.trxManager = trxManager;
+		this.ppOrderCandidatesDAO = ppOrderCandidatesDAO;
+		this.productPlanningsRepo = productPlanningsRepo;
+
+		this.result = new OrderGenerateResult();
+		this.planning2RecordShortTimeIndex = new ConcurrentHashMap<>();
+	}
+
+	@NonNull
+	public OrderGenerateResult createOrders(@NonNull final Stream<I_PP_Order_Candidate> candidates, @Nullable final Boolean isDocComplete)
+	{
+		final ImmutableList<PPOrderCandidateToAllocate> sortedCandidates = candidates
+				.filter(orderCandidate -> !orderCandidate.isProcessed())
+				.map(PPOrderCandidateToAllocate::of)
+				.sorted(Comparator.comparing(PPOrderCandidateToAllocate::getHeaderAggregationKey))
+				.collect(ImmutableList.toImmutableList());
+
+		if (sortedCandidates.isEmpty())
+		{
+			return result;
+		}
+
+		processPPOrderCandidates(sortedCandidates, isDocComplete);
+
+		return result;
+	}
+
+	private void processPPOrderCandidates(
+			@NonNull final ImmutableList<PPOrderCandidateToAllocate> ppOrderCandToAllocateSorted,
+			@Nullable final Boolean isDocComplete)
+	{
+		PPOrderAllocator allocator = null;
+		for (final PPOrderCandidateToAllocate ppOrderCandidateToAllocate : ppOrderCandToAllocateSorted)
+		{
+			if (allocator == null)
+			{
+				allocator = ppOrderAllocatorBuilderService.buildAllocator(ppOrderCandidateToAllocate);
+			}
+
+			boolean fullyAllocatedCandidate = false;
+			while (!fullyAllocatedCandidate)
+			{
+				fullyAllocatedCandidate = allocator.allocate(ppOrderCandidateToAllocate);
+
+				if (!fullyAllocatedCandidate)
+				{
+					createPPOrderInNewTrx(allocator, isDocComplete);
+
+					allocator = ppOrderAllocatorBuilderService.buildAllocator(ppOrderCandidateToAllocate);
+				}
+			}
+		}
+
+		if (allocator != null && allocator.getAllocatedQty().signum() > 0)
+		{
+			createPPOrderInNewTrx(allocator, isDocComplete);
+		}
+	}
+
+	public void createPPOrderInNewTrx(@NonNull final PPOrderAllocator allocator,@Nullable final Boolean completeDoc)
+	{
+		trxManager.runInNewTrx(() ->
```

Comment:
why a new trx? why not using current thread inherited trx?

Reply:
I was thinking about not loosing the existing progress => this way, if you have already created 10 PP_Orders and the 11th fails the 10 will still be around

Reply:
but I think the same effect would be achieved if we runInThreadInheritedTrx.
So it's shall be up to the caller to decide (e.g. de.metas.async.spi.IWorkpackageProcessor2#isRunInTransaction).

Btw, in case the BL generates 10 and fails at 11th, how will the recovery work?
e.g. "reprocess" will consider the previous 10 orders generated?

Reply:
yep, "reprocess" will consider the previous 10 orders generated as the progress is stored in `PP_OrderCandidate_PP_Order`.